### PR TITLE
Convert window shading flag to enum class

### PR DESCRIPTION
**Pull request by xuanluo113**
Pull request overview
---------------------
The PR includes the following refactoring efforts to window surface shading flags.
* Change `SurfWinShadingFlag` to ENUM class (e.g. `WinShadingType::IntShade`)
* Change `WindowShadingControl` to use the same ENUM classes as above
* Use `BITF_TEST` and Macros for checking window shading types.
* Clean up the logic of the `WindowShadingManager` function.
* Clean up the logic of the read, write, and EMS control usage of shading types.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by xuanluo113** [comment]
The branch is ready for review. @mjwitte Do you have the availability to review this? Thanks!

**Comment by mjwitte** [comment]
@xuanluo113 Yes, I will review. This branch has conflicts (likely from the merge I just did).

File: src/EnergyPlus/DataSurfaces.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh:92]: With hindsight, I wonder if these could be constexpr functions that still achieve the performance gains but allow more insight with IDEs.  Nothing to go back and change here for now, just something to think about before we spread this pattern around other places.

File: src/EnergyPlus/DataSurfaces.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh:196]: Yeah, I mean, this is just way better. :+1:

File: src/EnergyPlus/WindowManagerExteriorThermal.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WindowManagerExteriorThermal.cc:256]: I think the toughest part of this is going to remember which of these little worker functions should be used in each situation.

**Review by Myoldmopar — APPROVED** [review]
I have no problem here.  Converting a ton of `extern int const` into a proper enum, and then adding some preprocessor worker functions is all good.  As I said, it would be good to consider whether `constexpr` functions would work instead of the defines, but that's for later.  If CI is happy and @mjwitte is happy, I think this can go in.  I'm going to pull develop into it locally real quick and build/run one more time.

**Comment by Myoldmopar** [comment]
Built with latest develop and ran all tests serially and in parallel with no problem.  If no other issues arise with this, and nothing changes in develop, this can go in without more CI runs.

**Comment by Myoldmopar** [comment]
Develop moved so I went ahead and updated this branch. Building and running fine still.

File: src/EnergyPlus/WindowManagerExteriorThermal.cc
Comment (amirroth) [reply src/EnergyPlus/WindowManagerExteriorThermal.cc:256]: I think the shading logic needs to be refactored in general. But that can be done later.

File: src/EnergyPlus/DataSurfaces.hh
Comment (amirroth) [reply src/EnergyPlus/DataSurfaces.hh:92]: The `SHADE_FLAG` argument itself is not `constexpr`, so a `constexpr` function will not help or at least cannot be the entire solution. If you want something that is closer to the original code and maybe more transparent at the call site, you could use something like:

`#define BITF_TEST_ANY3(FLAG, B1, B2, B3) BITF_TEST_ANY(BITF(FLAG), BITF(B1) | BITF(B2) | BITF(B3))`

`if (BITF_TEST_ANY3(ShadeFlag, WinShadingType::IntBlind, WinShadingType::ExtBlind, WinShadingType::BGBlind))`

Is that better than:

`if (TEST_ANY_BLIND(ShadeFlag))`

? I don't know, is it?

File: src/EnergyPlus/DataSurfaces.hh
Comment (amirroth) [reply src/EnergyPlus/DataSurfaces.hh:196]: Mmmmmmmmm. Ok. Also, stick a pin in the fact that we want to provide a portable method for using enumerations in EMS and PythonPlugin so that we can do additional cleanup later.

File: src/EnergyPlus/DataSurfaces.hh
Comment (mjwitte) [inline src/EnergyPlus/DataSurfaces.hh:562]: Why is this changed from `int` to `Real64`?

File: src/EnergyPlus/DataSurfaces.cc
Comment (mjwitte) [inline src/EnergyPlus/DataSurfaces.cc:257]: So, it appears that Shading Types have been mixed in/consolidated with the Shade Status types. I'm not sure that's a good plan. While similar, `WSC_ST_InteriorShade` had a different meaning from `IntShadeOn`. I'll see how this plays out in the sections of code that use(d) these two sets of values.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc:1743]: In this function and most (all?) of the functions named `FigureDayltgCoeffs*`, `ShType` is the type of shade available to be applied to a particular window surface. These are preprocessing functions that setup coefficients to be used later in the actual timestep daylighting calcs.
1. ShType here should only be one of the nine values from the old "WindowShadingControl Shading Types" list. By using the new broader `WindowShadingControlType ` enum class are we opening ourselves up for potential problems?
2. Side note to myself: This is why multiple WindowShadingControl objects applied to the same window **MUST** have the same Shading Type!! This is currently enforced, but I had naïvely thought that we could allow different types of shades to be slapped on at any time. That might work for solar, but clearly based on these functions it won't work for daylighting.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc]: This could use a comment to explain what it is used for.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc]: Hmm, 'WinShadingIndex' implies more than just two states, 1=unshaded, 2=shaded. I guess the name of this doesn't matter as long at `findWinShadingIndex` has a comment to explain this is simply 1=unshaded, 2=shaded.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc:6792]: Hmm, I understand why, but it seems there must be an easier way to accomplish this. At least start with an outside if block using a new bit function `ifConditionallyOff` so the rest of this can be skipped if that's false?

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc:6867]: Is this also just checking if it's not in a condition state?

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 Some initial comments. Need to stop for now (at DaylightingManager.cc:6900.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc]: This seems like it should also have `(SurfWinWindowModelType(IWin) != WindowBSDFModel)`. I don't think IS=2 is ever valid for BSDF windows. Of course, fixing that here could introduce diffs.

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc:7947]: These variables names have always implied the wrong thing. They really should be named "HasShade" or "ShadeAvailable".

File: src/EnergyPlus/DaylightingManager.cc
Comment (mjwitte) [inline src/EnergyPlus/DaylightingManager.cc:9703]: Same here, I think it should have `(SurfWinWindowModelType(IWin) != WindowBSDFModel)`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:3569]: It is true that blinds cannot be used at the same time as a screen?

File: src/EnergyPlus/SolarShading.cc
Comment (mjwitte) [inline src/EnergyPlus/SolarShading.cc:6747]: What are the restrictions on using shading devices with BSDF windows? Can they ever have one?

**Review by mjwitte — COMMENTED** [review]
@xuanluo113 I've finished asking questions. I would like to see a unit test that tests the new macro (or future constexpr) functions for every possible shading flag value to be sure they all return the expected result. And to protect them against future breakage.

File: src/EnergyPlus/DataSurfaces.cc
Comment (xuanluo113) [reply src/EnergyPlus/DataSurfaces.cc:257]: @mjwitte Please see [this](https://github.com/NREL/EnergyPlus/pull/8435#commitcomment-46551726) thread for the explanation. You are right. Mixing the types and status was not an optimal design decision in the first place (I only reassured this when I was 80% done with this branch). But changing this may require some redesign of the logic and a lot more refactoring. It's more like a refactoring rather than a 10x effort. I think this could happen later.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc:3569]: I had the same question. I think according to the current logic, we cannot have `shadingflag == IntBlindOn && shadingflag == ExtScreenOn` (`shadingflag == 1 && shadingflag == 4`), because there is only one variable `shadingflag` to indicate which type of shades/blinds/screen/bgshade is on, and the `on` types are exclusive. Maybe we should also consider changing this logic.

File: src/EnergyPlus/DataSurfaces.hh
Comment (xuanluo113) [reply src/EnergyPlus/DataSurfaces.hh:562]: The EMS program reads

```
Set Shade_Status_None = 0.0 - 1.0,  !- Program Line 1
Set Shade_Status_Off = 0.0,  !- Program Line 2
Set Shade_Status_Interior_Shade_On = 1.0,  !- <none>
```

(I think) EnergyPlus originally has an implicit cast from real numbers to int. Here we explicitly "cast" real numbers to ENUMs. It's a temporary workaround we discuss with Edwin @Myoldmopar to make this work with ENUMs without transition changes.

File: src/EnergyPlus/DaylightingManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/DaylightingManager.cc:1743]: 1. Yes. Again it's the decoupling of types and status stuff. I can either (1) move the `WindowShadingControlType` to another enum class, or (2) move the `ShadeOff` and `ShadeCondOff` out of the current enums and change the entire logic, or (3) add comments... 
2. Exactly. Then it kinda aligned with what I replied above. I think it's totally weird to me. Maybe it's another reason to fix the logic in future branch/release.

File: src/EnergyPlus/DaylightingManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/DaylightingManager.cc:6792]: You can only access this branch of code when `ifConditionallyOn` or `SwitchableGlazing`, enforced by
```
if (NOT_SHADED(SurfWinShadingFlag(IWin)) || ANY_SHADE_SCREEN(SurfWinShadingFlag(IWin)) || ANY_BLIND(SurfWinShadingFlag(IWin))) {
    continueOuterLoop = false;
    continue;
}
```
(originally `if (SurfWinShadingFlag(IWin) < 10 && SurfWinShadingFlag(IWin) != SwitchableGlazing) continue;`)

File: src/EnergyPlus/DaylightingManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/DaylightingManager.cc:6867]: Again, when `ifConditionallyOn` or `SwitchableGlazing`.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc
Comment (mjwitte) [reply src/EnergyPlus/HeatBalanceSurfaceManager.cc:3569]: Yes, of course. I was remembering the storm window flag, which can be independent of shading status.

File: src/EnergyPlus/DaylightingManager.cc
Comment (xuanluo113) [reply src/EnergyPlus/DaylightingManager.cc:6867]: (honestly I don't get the `ifConditionallyOn` or `SwitchableGlazing` logic that happens in a lot of places)

File: src/EnergyPlus/SolarShading.cc
Comment (xuanluo113) [reply src/EnergyPlus/SolarShading.cc:6747]: The flag can be set as `ExtShade` or `IntShade` . https://github.com/NREL/EnergyPlus/blob/e652b59caee1ab6cafbb56973eff96275379e3d2/src/EnergyPlus/SolarShading.cc#L8674

**Comment by xuanluo113** [comment]
@mjwitte Thanks for the review.  I've replied to those issues that currently I don't have a concrete solution or need further considerations or discussion to fix. I'll work on fixing the rest of the comments which I marked a thumb there.

**Comment by mjwitte** [comment]
@xuanluo113 Thanks for the clarifications. This is an impressive piece of work. Sorting out all of the old compound if logic into meaningful groupings is quite challenging. It would make my head spin.

**Review by mjwitte — APPROVED** [review]
@Myoldmopar I'm ok with these changes. @xuanluo113 has a little bit of cleanup to do and then I'll let you make the final review/merge.

**Comment by xuanluo113** [comment]
@mjwitte @Myoldmopar Comments are addressed in this branch. Hope it's ready to get in.

**Comment by mjwitte** [comment]
@xuanluo113 Please pull in develop one last time, then I'll merge if CI is still green.

**Comment by xuanluo113** [comment]
@mjwitte Done. CI looks okay.

**Comment by mjwitte** [comment]
Now that 9.5 IOFreeze is tagged, I'm merging this.

**Pull request merged**
